packages/embedder/ai_models/RecursiveCharacterTextSplitter.ts

```diff
@@ -0,0 +1,117 @@
+// mostly borrowed from langchain
+export class RecursiveCharacterTextSplitter {
+  separators = ['\n\n', '\n', ' ', '']
+  keepSeparator = true
+  chunkSize: number
+  chunkOverlap: number
+  constructor({chunkSize, chunkOverlap}: {chunkSize: number; chunkOverlap: number}) {
+    this.chunkSize = chunkSize
+    this.chunkOverlap = chunkOverlap || 0
+  }
+  private splitOnSeparator(text: string, separator: string): string[] {
+    let splits: string[] = []
+    if (separator) {
+      if (this.keepSeparator) {
+        const regexEscapedSeparator = separator.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
+        splits = text.split(new RegExp(`(?=${regexEscapedSeparator})`))
+      } else {
+        splits = text.split(separator)
+      }
+    } else {
+      splits = text.split('')
+    }
+    return splits.filter((s) => s !== '')
+  }
+
+  private joinDocs(docs: string[], separator: string): string | null {
+    const text = docs.join(separator).trim()
+    return text === '' ? null : text
+  }
+
+  private mergeSplits(splits: string[], separator: string) {
+    const docs: string[] = []
+    const currentDoc: string[] = []
+    let total = 0
+    for (const d of splits) {
+      const _len = d.length
+      if (total + _len + (currentDoc.length > 0 ? separator.length : 0) > this.chunkSize) {
+        if (total > this.chunkSize) {
```

Comment:
you know, i'm not sure! i just blindly copied most of this from https://github.com/langchain-ai/langchainjs/blob/2c31e559558d9c2aacc4c1f01a0f7dd29e933ac0/langchain/src/text_splitter.ts#L191

which ported it from python
https://github.com/langchain-ai/langchain/blob/5d220975fc563a92f41aeb0907e8c3819da073f5/libs/text-splitters/langchain_text_splitters/base.py#L117-L118

since the JS matches the python logic i'm assuming it's probably good, but i didn't test it!